contracts/pm/TicketBroker.sol

```diff
@@ -114,6 +127,67 @@ contract TicketBroker {
         );
     }
 
+    function unlock() public {
+        Sender storage sender = senders[msg.sender];
+
+        require(
+            sender.deposit > 0 || sender.penaltyEscrow > 0,
+            "sender deposit and penalty escrow are zero"
+        );
+        require(!_isUnlockInProgress(sender), "unlock already initiated");
+
+        sender.withdrawBlock = block.number + unlockPeriod;
+
+        emit Unlock(msg.sender);
+    }
+
+    function cancelUnlock() public {
+        Sender storage sender = senders[msg.sender];
+
+        _cancelUnlock(sender, msg.sender);
+    }
+
+    function withdraw() public {
+        Sender storage sender = senders[msg.sender];
+
+        require(
+            sender.deposit > 0 || sender.penaltyEscrow > 0,
+            "sender deposit and penalty escrow are zero"
+        );
+        require(
+            _isUnlockInProgress(sender) && block.number >= sender.withdrawBlock, 
+            "account is locked"
+        );
+
+        uint256 withdrawalAmount = sender.deposit + sender.penaltyEscrow;
+        sender.deposit = 0;
+        sender.penaltyEscrow = 0;
+
+        withdrawTransfer(msg.sender, withdrawalAmount);
+
+        emit Withdrawal(msg.sender, withdrawalAmount);
+    }
+
+    function isUnlockInProgress(address _sender) public view returns (bool) {
+        Sender memory sender = senders[_sender];
+        return _isUnlockInProgress(sender);
+    }
+
+    function _cancelUnlock(Sender storage _sender, address _senderAddress) internal {
+        require(_isUnlockInProgress(_sender), "no unlock request in progress");
```

Comment:
The `processDeposit` and `processPenaltyEscrow` modifiers already have a check using `_isUnlockInProgress()` so we could move this require statement into `cancelUnlock()`

Reply:
here's why I thought the current approach makes sense:

- this `require` is here to make sure we don't emit an event unless it's really interesting
- I still want to keep it here in case we use this function in other places and don't wrap it with an if or require statement
- I added the if statements in the modifiers to prevent them from failing because of this error

If I do what you suggest we leave this function open to misuse such that the event might be emitted even if `withdrawBlock` was already set to zero.

wdyt?

Reply:
My suggestion was originally motivated by a desire to avoid 2 `_isUnlockInProgress()` checks, but your points make sense. I am ok with repeating a check if it means safer usage of `_cancelUnlock()`

Reply:
awesome!